### PR TITLE
refactor: Extract discovery of rule violations from SoraldAbstractRepair

### PR DESCRIPTION
**Pull request by slarse**
This is step 1 of the targeted repair implementation (#225).

This refactor extracts the discovery of rule violations from the SoraldAbstractRepair class and pulls it up to an earlier level of the execution. It makes it much easier to replace the rule violations with something constructed from violation IDs for the targeted repair. See further comments in the code.

### CODE REVIEW

File: src/main/java/sorald/Repair.java [hunk 1]
```diff
@@ -65,7 +69,13 @@ public void repair() {
...
             final Path outputDir = inOutPaths.getRight();
 
-            SoraldAbstractProcessor<?> processor = createProcessor(ruleKey);
+            Set<RuleViolation> ruleViolations =
+                    ProjectScanner.scanProject(
+                            inputDir.toFile(),
+                            FileUtils.getClosestDirectory(inputDir.toFile()),
+                            Checks.getCheckInstance(Integer.toString(ruleKey)));
+            SoraldAbstractProcessor<?> processor =
+                    createProcessor(ruleKey).setRuleViolations(ruleViolations);
```
Comment (slarse) [inline src/main/java/sorald/Repair.java:78]: We now set rule violations here, at the very top of the repair process. This makes it easy to swap these currently discovered violations for something we construct ourselves from parsing rule IDs for targeted repair.

File: src/main/java/sorald/processor/SoraldAbstractProcessor.java [hunk 2]
```diff
@@ -31,49 +22,8 @@
...
-                try (Stream<Path> walk = Files.walk(Paths.get(node.getRootPath()))) {
-                    filesToScan.addAll(
-                            walk.map(x -> x.toFile().getAbsolutePath())
-                                    .filter(f -> f.endsWith(Constants.JAVA_EXT))
-                                    .collect(Collectors.toList()));
-                } catch (IOException e) {
-                    e.printStackTrace();
-                }
-            }
-        }
```
Comment (slarse) [inline src/main/java/sorald/processor/SoraldAbstractProcessor.java:74]: This method is not actually necessary. It's more efficient to scan the entire project for rule violations before splitting it into segments, than to scan the segments one-by-one. It doesn't matter that when repairing segment A there are also violations for segment B in the violations set of the processor, as they won't match anyway.

File: src/main/java/sorald/Repair.java (hunk 1 above)
Comment (fermadeiral) [reply src/main/java/sorald/Repair.java:78]: Nice!

File: src/main/java/sorald/processor/SoraldAbstractProcessor.java (hunk 2 above)
Comment (fermadeiral) [reply src/main/java/sorald/processor/SoraldAbstractProcessor.java:74]: This code was also duplicate, it's nice that it's gone!

**Comment by fermadeiral** [comment]
Thanks for the refactoring!

**Pull request merged**